OAK-11935 : updating aws sdk from 1.x to 2.x

rishabhdaim opened the pull request.

rishabhdaim committed:
OAK-11935 : updating aws sdk from 1.x to 2.x

reschke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/DataEncryption.java:
```diff
@@ -0,0 +1,25 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
should this be marked internal?

rishabhdaim replied:
Yes

reschke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -421,15 +414,20 @@ public InputStream read(DataIdentifier identifier)
+                            .build()),
+                    ResponseTransformer.toInputStream());
+
+            LOG.debug("[{}] read took [{}]ms", identifier, System.currentTimeMillis() - start);
```
Stopwatch?

rishabhdaim replied:
They were written using `System.currentTimeMillis`, so we used that to keep the symmetry.

reschke replied:
you can get millis from StopWatch as well. but yes, just a nit

reschke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -1276,22 +1287,24 @@ private void renameKeys() throws DataStoreException {
             }
-            LOG.info("Renamed [{}] keys, time taken [{}]sec", count,
-                ((System.currentTimeMillis() - startTime) / 1000));
+            LOG.info("Renamed [{}] keys, time taken [{}]sec", count, ((System.currentTimeMillis() - startTime) / 1000));
```
Stopwatch?

rishabhdaim replied:
same as above.

reschke reviewed (commented):
nits

rishabhdaim committed:
OAK-11935 : added more unit cases

rishabhdaim committed:
OAK-11935 : added unit cases for S3RequestDecorator

rishabhdaim committed:
OAK-11935 : fixed issue when running ITs with different encryption mode

rishabhdaim committed:
OAK-11935 : removed unused imports

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:
```diff
@@ -101,109 +117,96 @@ public GetObjectRequest decorate(final GetObjectRequest request) {
-                /*Set*/
-                request.withSSEAwsKeyManagementParams(sseParams);
+                builder.serverSideEncryption(ServerSideEncryption.AWS_KMS);
+                if (sseKmsKey != null) {
```
what happens when sseKmsKey == null?

rishabhdaim replied:
It will work fine, we can run it without `kmeKey`.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:
```diff
@@ -101,109 +117,96 @@ public GetObjectRequest decorate(final GetObjectRequest request) {
             case SSE_C:
-                request.withSSECustomerKey(sseCustomerKey);
+                builder.sseCustomerAlgorithm(ServerSideEncryption.AES256.toString());
+                if (sseCustomerKey != null) {
```
what happens when sseCustomerKey == null?
Will any request work?

rishabhdaim replied:
No, this won't work. But didn't update the logic to avoid any unnecessary change.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:
```diff
@@ -101,109 +117,96 @@ public GetObjectRequest decorate(final GetObjectRequest request) {
-                metadata.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm());
-                request.withSSEAwsKeyManagementParams(sseParams);
+                builder.serverSideEncryption(ServerSideEncryption.AWS_KMS);
+                if (sseKmsKey != null) {
```
Q: same as above

rishabhdaim replied:
it would work.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:
```diff
@@ -101,109 +117,96 @@ public GetObjectRequest decorate(final GetObjectRequest request) {
-                request.withSourceSSECustomerKey(sseCustomerKey).withDestinationSSECustomerKey(sseCustomerKey);
+                builder.sseCustomerAlgorithm(ServerSideEncryption.AES256.toString());
+                builder.copySourceSSECustomerAlgorithm(ServerSideEncryption.AES256.toString());
+                if (sseCustomerKey != null) {
```
Q: same as above?

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:
```diff
@@ -101,109 +117,96 @@ public GetObjectRequest decorate(final GetObjectRequest request) {
             case SSE_C:
-                request.withSSECustomerKey(sseCustomerKey);
+                builder.sseCustomerAlgorithm(ServerSideEncryption.AES256.toString());
+                if (sseCustomerKey != null) {
```
Q: Same as above.
Point is if it does not work why not throw in init itself.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/Utils.java:
```diff
@@ -67,70 +102,84 @@ public final class Utils {
+        // region is mandatory even with endpointOverride
+        builder.region(Region.of(region));
+        // to enable cross region bucket access
+        builder.crossRegionAccessEnabled(true);
```
Do we need another property for this?

rishabhdaim replied:
We can have a property with a default to `false` to have previous behaviour.

rishabhdaim replied:
added in https://github.com/apache/jackrabbit-oak/pull/2558/commits/b02f5303dbe45cb52189a081653e9f2d856a9a1a

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/Utils.java:
```diff
@@ -67,70 +102,84 @@ public final class Utils {
-        s3service.setEndpoint(endpoint);
-        LOG.info("S3 service endpoint [{}] ", endpoint);
-        return s3service;
+    public static S3Client openService(final Properties prop, boolean accReq) {
```
Why is accelerate request not in properties and a separate parameter?

rishabhdaim replied:
this can be optimized by referring it from `properties`.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/Utils.java:
```diff
@@ -216,54 +290,384 @@ public static Properties readConfig(String fileName) throws IOException {
+                                    .key(key)
+                                    .build()));
+
+                }
```
Won't this work?
```suggestion
        for (ListObjectsV2Response listRes : listResponses) {
            List<ObjectIdentifier> deleteList = new ArrayList<>();
            for (S3Object s3Obj : listRes.contents()) {
                deleteList.add(ObjectIdentifier.builder().key(s3Obj.key()).build());
            }

            if (!deleteList.isEmpty()) {
                RemoteStorageMode mode = getMode(props);
                if (mode == RemoteStorageMode.S3) {
                    s3service.deleteObjects(delReq ->
                            delReq.bucket(bucket)
                                    .delete(delObj ->
                                            delObj.objects(deleteList)));
                } else {
                    // Delete objects one by one
                    deleteList.forEach(obj ->
                        s3service.deleteObject(delReq ->
                            delReq.bucket(bucket).key(obj.key())));

                }
            }
```

rishabhdaim replied:
addressed in https://github.com/apache/jackrabbit-oak/pull/2558/commits/922ecdfd265cb15a7e140465c8b29c604f1a4ccb

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/Utils.java:
```diff
@@ -146,46 +195,71 @@ public static AmazonS3Client openService(final Properties prop) {
+
+    public static void deleteBucketObjects(final String bucket, final Properties props, final S3Client s3service) {
+        ListObjectsV2Iterable listResponses = s3service.listObjectsV2Paginator(builder -> builder.bucket(bucket).build());
+        deleteBucketObjects(bucket, props, s3service, listResponses);
```
Don't see a reason why the logic cannot be inlined, is anybody calling this method directly?

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/Utils.java:
```diff
@@ -216,54 +290,384 @@ public static Properties readConfig(String fileName) throws IOException {
+     * @return the constructed S3 endpoint URI
+     */
+    @NotNull
+    static URI getEndPointUri(Properties prop, boolean accReq, String region) {
```
as above why can't accelerate request not be a part of props?

rishabhdaim replied:
This is not entirely based on properties value, we also need to check whether acceleration is enabled on given bucket.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -236,18 +223,17 @@ public void init() throws DataStoreException {
 
             String enablePresignedAccelerationStr = properties.getProperty(S3Constants.PRESIGNED_URI_ENABLE_ACCELERATION);
-            setBinaryTransferAccelerationEnabled(enablePresignedAccelerationStr != null && "true".equals(enablePresignedAccelerationStr));
+            setBinaryTransferAccelerationEnabled("true".equals(enablePresignedAccelerationStr));
```
Why can't this block be moved at the start and accelerate prop added and then rest of init takes place as normal, we will avoid opening the diff service object again.

rishabhdaim replied:
It was written this way earlier, so I didn't make any changes to it.
But I can move the logic upwards and make it more streamlined.

rishabhdaim replied:
Had to re-look, this it is not possible we because we an `s3Client` to verify whether `accelerate` is enabled on this `bucket` or not.

If enabled, only then can we start creating accelerated presigned URLs.

```
void setBinaryTransferAccelerationEnabled(boolean enabled) {
        if (enabled) {
            // verify acceleration is enabled on the bucket
            GetBucketAccelerateConfigurationResponse accelerateConfig = s3Client.getBucketAccelerateConfiguration(b -> b.bucket(bucket).build());
            if (Objects.equals(BucketAccelerateStatus.ENABLED, accelerateConfig.status())) {
                // If transfer acceleration is enabled for presigned URIs, we need a separate AmazonS3Client
                // instance with the acceleration mode enabled, because we don't want the requests from the
                // data store itself to S3 to use acceleration
                s3PresignService = Utils.createPresigner(Utils.openService(properties, true), properties);
                LOG.info("S3 Transfer Acceleration enabled for presigned URIs.");

            } else {
                LOG.warn("S3 Transfer Acceleration is not enabled on the bucket {}. Will create normal, non-accelerated presigned URIs. To enable set {}",
                    bucket, S3Constants.PRESIGNED_URI_ENABLE_ACCELERATION);
            }
        } else {
            s3PresignService = Utils.createPresigner(s3Client, properties);
        }
    }
```

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -316,57 +306,72 @@ void setBinaryTransferAccelerationEnabled(boolean enabled) {
+                                .metadataDirective(MetadataDirective.REPLACE)
+                                .build();
+                LOG.warn("Object MetaData before copy: {}", headMeta.metadata());
                 if (Objects.equals(RemoteStorageMode.S3, properties.get(S3Constants.MODE))) {
```
MetadataDirective.COPY cannot be used here? I think then this block is not needed.

rishabhdaim replied:
No, this is required to update metadata and reset creation time. Without this, the ITs started failing.
Also, COPY is the default behaviour.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -391,21 +395,10 @@ public boolean exists(DataIdentifier identifier) throws DataStoreException {
-            }
-            throw new DataStoreException(
-                "Error occured to getObjectMetadata for key [" + identifier.toString() + "]", e);
+            return Utils.objectExists(s3Client, bucket, key, s3ReqDecorator);
```
It is a little problematic that a core method has been moved to th Utils which was originally handling connection setup. For tests if required should duplicate it to the DataStoreUtils to be used.

rishabhdaim replied:
This was added to avoid duplicate code.
It was becoming cumbersome for me to update the same request at different places, and also led to many errors once forgot.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -469,11 +463,20 @@ public void deleteRecord(DataIdentifier identifier)
+            if (s3AsyncClient != null) {
+                s3AsyncClient.close();
+            }
+            s3Client.close();
```
As the code stands we should close the s3PresignService or the s3Utils backing it, because it creates a new s3Client in some cases but with the suggested changes it may be necessary.

rishabhdaim replied:
Sorry, I Didn't get this?

amit-jain replied:
`            s3PresignService = Utils.createPresigner(Utils.openService(properties, true), properties);`
with this a new s3 client object is being created but not closed but I see now the s3Presigner close has been added

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -496,39 +499,41 @@ public void setProperties(Properties properties) {
         ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
 
+        // Executor required to handle reading from the InputStream on a separate thread so the main upload is not blocked.
+        final ExecutorService executor = Executors.newSingleThreadExecutor();
```
Can we setup a pool at the beginning and close in the close()?

rishabhdaim replied:
1. It was written in this way since the beginning, so I didn't update the logic.
2. This thread pool is only used once, and in this method, so I don't think we should declare it at the global/class level.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -624,29 +633,13 @@ public void deleteAllMetadataRecords(String prefix) {
+                    .bucket(bucket)
+                    .prefix(addMetaKeyPrefix(prefix))
+                    .build());
+            Utils.deleteBucketObjects(bucket, properties, s3Client, metaList);
```
same case as Utils.objectExists, for other invocations as well

rishabhdaim replied:
Same reason, had to avoid duplicate code.

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -1075,80 +1110,52 @@ private URI createPresignedURI(DataIdentifier identifier,
-
-            return endOfData();
-        }
+        public RecordsIterator(Function<S3Object, T> transformer) {
```
looks to me it is completelty unnecessary now with the new sdk.
Maybe we can add something like below and in both the use cases invoke it the coreect transformer.

`
   ListObjectsV2Iterable listing = s3Client.listObjectsV2Paginator(builder -> {
        builder.bucket(bucket);
        if (properties.containsKey(S3Constants.MAX_KEYS)) {
            builder.maxKeys(Integer.valueOf(properties.getProperty(S3Constants.MAX_KEYS)));
        }
    });
 
    return IteratorUtils.transform(
        IteratorUtils.filter(
            listing.contents().iterator(),
            s3Object -> !s3Object.key().startsWith(META_KEY_PREFIX)
        ),
        transformer
    );
`

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -1346,34 +1359,51 @@ private static String getIdentifierName(String key) {
         return key.substring(0, 4) + key.substring(5);
     }
 
+    private void setRemoteStorageMode() {
```
misplaced? Should it not be in the Utils?

rishabhdaim replied:
Only used in this class, so didn't move it to Utils.
But, I agree this can be moved, so strong opinion here.

amit-jain commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/fixture/DataStoreUtils.java:
```diff
@@ -112,40 +108,11 @@ public static void cleanup(File storeDir, Map<String, ?> config, String bucket)
-        }
-        tmx.shutdownNow();
-        s3service.shutdown();
+        Utils.deleteBucketAndAbortMultipartUploads(bucket, date, props);
```
I think having a core crud kind of method in Utils is not a good place

rishabhdaim committed:
OAK-11935 : incorporated review comments

rishabhdaim commented:
@amit-jain I have addressed your comments in https://github.com/apache/jackrabbit-oak/pull/2558/commits/793126f7da00dc8a07faf332894203539ca4fb05.
please have re-look.

Regarding moving certain methods to `Utils` class, in case you are not ok with that, we need to discuss suitable place for them.

reschke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/DataEncryption.java:
```diff
@@ -0,0 +1,28 @@
+ * Enum to indicate S3 encryption mode
+ *
+ */
+@Internal
```
Sry. no. I saw a package-info where non is.

reschke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -235,19 +226,15 @@ public void init() throws DataStoreException {
+            LOG.debug("S3 Backend initialized in [{}] ms", (System.currentTimeMillis() - startTime.getTime()));
         } catch (Exception e) {
-            LOG.error("Error ", e);
+            LOG.error("Error while initializing S3 Backend {}", e.getMessage());
```
ok to loose the stack trace?

rishabhdaim replied:
added in https://github.com/apache/jackrabbit-oak/commit/a52bd37b68f9e8f38967cf42511e374603a16f70

mbaedke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -1022,134 +1026,102 @@ record = new S3DataRecord(
+
+                    presignedURL = s3PresignService.presignPutObject(builder -> builder
+                            .signatureDuration(Duration.between(
+                                    java.time.Instant.now(),           // current time
```
No need to qualify.

rishabhdaim replied:
added in https://github.com/apache/jackrabbit-oak/commit/a52bd37b68f9e8f38967cf42511e374603a16f70

mbaedke commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -1276,22 +1253,24 @@ private void renameKeys() throws DataStoreException {
-                                startIndex, endIndex)));
-                        DeleteObjectsResult dobjs = s3service.deleteObjects(delObjsReq);
+                        int finalStartIndex = startIndex;
+                        int finalEndIndex = endIndex;
```
What are these good for?

rishabhdaim replied:
added in https://github.com/apache/jackrabbit-oak/pull/2558/commits/a52bd37b68f9e8f38967cf42511e374603a16f70

mbaedke reviewed (approved):
Other than nits lgtm. For what it's worth :)

rishabhdaim committed:
OAK-11935 : added review comments

rishabhdaim committed:
OAK-11935 : added properties for cross Region access

rishabhdaim committed:
OAK-11935 : removed additional list for removing keys in case of GCP mode

rishabhdaim committed:
OAK-11935 : moved out CRUD operations from Utils to S3CrudHelper

rishabhdaim committed:
OAK-11935 : added unit cases for waitForBucket method

rishabhdaim committed:
OAK-11935 : make total attemps and delay configurable to use lower values in unit cases

rishabhdaim committed:
OAK-11935 : remaned S3CrudHelper to S3BackendHelper

amit-jain commented on oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:
```diff
@@ -516,19 +509,34 @@ public void addMetadataRecord(final InputStream input, final String name) throws
             if (contextClassLoader != null) {
                 Thread.currentThread().setContextClassLoader(contextClassLoader);
             }
+            executor.shutdown();
```
Should preferably use https://github.com/apache/jackrabbit-oak/blob/trunk/oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/concurrent/ExecutorCloser.java here

amit-jain reviewed (approved):
apart from some minor things, lgtm!
